### PR TITLE
pollution fixes

### LINKED ISSUE #1421: GTM + Xaero's Maps = logspam
### Checked for existing issues

- [X] I have checked for existing issues, and have found none.

### Tested latest version

- [X] I have checked that this occurs on the latest version.

### GregTech CEu Version

1.3.0

### Recipe Viewer Installed

None

### Environment

Singleplayer

### Cross-Mod Interaction

Yes

### Other Installed Mods

Xaero's World Map, Xaero's Minimap

### Expected Behavior

No log spam

### Actual Behavior

Log spam :(

### Steps to Reproduce

1. Load world
2. Log spam

### Additional Information

exception:

`[19:10:17] [Render thread/ERROR] [xa.ma.WorldMap/]: Error calculating block tint for block state Block{gtceu:pyrochlore_ore}! Total: 1
java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 2
	at it.unimi.dsi.fastutil.ints.IntArrayList.getInt(IntArrayList.java:342) ~[fastutil-8.5.9.jar%2394!/:?] {}
	at com.gregtechceu.gtceu.api.data.chemical.material.Material.getMaterialARGB(Material.java:379) ~[gtceu-1.20.1-1.3.0.jar%23131!/:?] {re:classloading}
	at com.gregtechceu.gtceu.api.data.chemical.material.Material.getLayerARGB(Material.java:359) ~[gtceu-1.20.1-1.3.0.jar%23131!/:?] {re:classloading}
	at com.gregtechceu.gtceu.api.block.MaterialBlock.lambda$tintedColor$0(MaterialBlock.java:88) ~[gtceu-1.20.1-1.3.0.jar%23131!/:?] {re:classloading}
	at net.minecraft.client.color.block.BlockColors.m_92577_(BlockColors.java:100) ~[client-1.20.1-20230612.114412-srg.jar%23134!/:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:mixin:A,pl:runtimedistcleaner:A}
	at xaero.map.biome.BlockTintProvider.getBiomeColor(BlockTintProvider.java:94) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.MapPixel.getPixelColours(MapPixel.java:78) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.Overlay.getPixelColour(Overlay.java:50) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.MapPixel.getPixelColours(MapPixel.java:107) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.MapBlock.getPixelColour(MapBlock.java:55) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.MapTileChunk.updateBuffers(MapTileChunk.java:97) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.region.texture.LeafRegionTexture.preUpload(LeafRegionTexture.java:47) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.MapProcessor.onRenderProcess(MapProcessor.java:476) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.events.ClientEvents.handleRenderTick(ClientEvents.java:86) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.events.ClientEventsForge.renderTick(ClientEventsForge.java:32) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading}
	at xaero.map.events.__ClientEventsForge_renderTick_RenderTickEvent.invoke(.dynamic) ~[XaerosWorldMap_1.38.8_Forge_1.20.jar%23133!/:1.38.8] {re:classloading,pl:eventbus:B}
	at net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:73) ~[eventbus-6.0.5.jar%23109!/:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:315) ~[eventbus-6.0.5.jar%23109!/:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:296) ~[eventbus-6.0.5.jar%23109!/:?] {}
	at net.minecraftforge.event.ForgeEventFactory.onRenderTickEnd(ForgeEventFactory.java:910) ~[forge-1.20.1-47.2.21-universal.jar%23139!/:?] {re:classloading}
	at net.minecraft.client.Minecraft.m_91383_(Minecraft.java:1148) ~[client-1.20.1-20230612.114412-srg.jar%23134!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,xf:fml:xaerominimap:xaero_minecraftclient,xf:fml:xaeroworldmap:xaero_wm_minecraftclient,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,xf:fml:xaerominimap:xaero_minecraftclient,xf:fml:xaeroworldmap:xaero_wm_minecraftclient,pl:mixin:APP:configuration.mixins.json:MinecraftMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:718) ~[client-1.20.1-20230612.114412-srg.jar%23134!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,xf:fml:xaerominimap:xaero_minecraftclient,xf:fml:xaeroworldmap:xaero_wm_minecraftclient,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,xf:fml:xaerominimap:xaero_minecraftclient,xf:fml:xaeroworldmap:xaero_wm_minecraftclient,pl:mixin:APP:configuration.mixins.json:MinecraftMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.main.Main.main(Main.java:218) ~[minecraft-1.20.1-client.jar:?] {re:classloading,pl:runtimedistcleaner:A}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {}
	at net.minecraftforge.fml.loading.targets.CommonLaunchHandler.runTarget(CommonLaunchHandler.java:111) ~[fmlloader-1.20.1-47.2.21.jar:?] {}
	at net.minecraftforge.fml.loading.targets.CommonLaunchHandler.clientService(CommonLaunchHandler.java:99) ~[fmlloader-1.20.1-47.2.21.jar:?] {}
	at net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$makeService$0(CommonClientLaunchHandler.java:25) ~[fmlloader-1.20.1-47.2.21.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:30) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:108) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:78) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:141) ~[bootstraplauncher-1.1.2.jar:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {}
	at io.github.zekerzhayard.forgewrapper.installer.Main.main(Main.java:67) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {}
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:243) ~[?:?] {}
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:278) ~[?:?] {}
	at org.multimc.EntryPoint.listen(EntryPoint.java:143) ~[?:?] {}
	at org.multimc.EntryPoint.main(EntryPoint.java:34) ~[?:?] {}`

### PR DESCRIPTION
**Pull request by screret**
## What
removes the client lag caused by pollution by reducing particle count and instead making larger particles.
also makes singleblock generators output very low amounts of pollution and makes the strength be stored as a float so smaller amounts can be output at a time.

## Implementation Details
new particle type that's like the basalt delta particle but colorable

## Outcome
less laggy environmental hazards

### CODE REVIEW

**Comment by 300f20t** [comment]
It would be nice if this fix increased the maximum pollution limit, as it is too small.

**Comment by screret** [comment]
> It would be nice if this fix increased the maximum pollution limit, as it is too small.

what?

**Comment by 300f20t** [comment]
The pollution is too large and when transferring from the old version of GTm it becomes maximum and because of this it becomes impossible to play after 2-3 minutes of being on the base. Maybe I should have opened the issue.

**Comment by screret** [comment]
that's... what this fixes.

**Comment by 300f20t** [comment]
The fact that the maximum pm is 1000 is a pretty bad decision. In GT5, this value was several million.

**Comment by screret** [comment]
it's not a maximum, it's the value at which it'll start spreading to chunks around it.

**Comment by screret** [comment]
also this is not that close to GT5 pollution and smaller amounts are much more deadly

**Comment by 300f20t** [comment]
It behaves very strangely and does not rise above 1000, but does not spread to other chunks.

**Comment by screret** [comment]
have you tried the attached development build?
[build output.zip](https://github.com/GregTechCEu/GregTech-Modern/actions/runs/9519977572/artifacts/1603007866)

**Comment by 300f20t** [comment]
I didn't want to create a issue because I couldn't reproduce the bug.

**Comment by 300f20t** [comment]
image

**Comment by 300f20t** [comment]
I'm not sure, but it seems to have worked.

**Comment by 300f20t** [comment]
I couldn't find any information about the pollution other than your commits on Github. Does the modification include any means of removing pollution?

**Comment by screret** [comment]
that already exists, it's called the Air Scrubber.

File: src/main/java/com/gregtechceu/gtceu/client/EnvironmentalHazardClientHandler.java [hunk 1]
```diff
@@ -42,119 +43,193 @@ private EnvironmentalHazardClientHandler() {
...
         }
     }
 
     public void removeHazardZone(ChunkPos pos) {
         this.hazardZones.remove(pos);
+
+        for (int y = Minecraft.getInstance().level.getMinSection(); y <
+                Minecraft.getInstance().level.getMaxSection(); ++y) {
+            Minecraft.getInstance().levelRenderer.setSectionDirtyWithNeighbors(pos.x, y, pos.z);
+        }
```
Comment (mikerooni) [inline src/main/java/com/gregtechceu/gtceu/client/EnvironmentalHazardClientHandler.java]: this part should probably be its own method as it's repeated a couple times above

File: src/main/java/com/gregtechceu/gtceu/client/EnvironmentalHazardClientHandler.java [hunk 2]
```diff
@@ -42,119 +43,193 @@ private EnvironmentalHazardClientHandler() {
...
-        return colorize(color, zone.strength(), zone.condition().color, FOLIAGE_COLORS);
+        var entry = chunkColorCache.get(pos);
+        if (entry != null &&
+                (entry.firstFloat() > zone.strength() + 0.5f || entry.firstFloat() < zone.strength() - 0.5f)) {
+            return entry.valueInt();
+        }
+
+        color = colorize(color, zone.strength(), zone.condition().color);
+        chunkColorCache.put(pos, FloatIntPair.of(zone.strength(), color));
+        return color;
```
Comment (mikerooni) [inline src/main/java/com/gregtechceu/gtceu/client/EnvironmentalHazardClientHandler.java:170]: the same goes for this part

**Review by mikerooni — CHANGES_REQUESTED** [review]

**Review by mikerooni — APPROVED** [review]

**Pull request merged**